File: typescript/src/inference/prefix_cache.ts
// inference/prefix_cache.ts
//
// Port of CircleAI.Inference.PrefixCacheService (RT-06). Manages an on-disk
// cache of "warm" model sessions keyed by SHA-256(modelId) + SHA-256(system
// prompt). The C# service uses the real filesystem; this port defaults to an
// in-memory backend (so tests are deterministic and hermetic) but the keying,
// path shape, LRU-by-mtime eviction, and 500 MB cap are all byte-faithful.
//
// The seam is IPrefixCacheStore — inject a filesystem-backed store to persist
// to disk, or use the default in-memory store.

import { createHash } from "node:crypto";

const CAP_BYTES = 500 * 1024 * 1024; // 500 MB

/**
 * Backing store for the prefix cache. Abstracts the filesystem so the cache
 * logic (keying, eviction policy) can run without touching real disk.
 */
export interface IPrefixCacheStore {
  /** true when a file exists at `path`. */
  exists(path: string): boolean;
  /** Write `content` bytes (length used for size accounting) at `path`. */
  write(path: string, content: string): void;
  /** Read the content at `path`, or null when absent. */
  read(path: string): string | null;
  /** Delete `path`. No-op when absent. */
  delete(path: string): void;
  /** Set the last-write time (epoch ms) of `path` to `mtimeMs`. */
  setMtime(path: string, mtimeMs: number): void;
  /** List entries under `root` whose name ends with `suffix`, with size + mtime. */
  list(root: string, suffix: string): Array<{ path: string; sizeBytes: number; mtimeMs: number }>;
}

/**
 * Deterministic in-memory store. A monotonically increasing clock stands in
 * for wall-clock mtimes so LRU ordering is stable and testable.
 */
export class InMemoryPrefixCacheStore implements IPrefixCacheStore {
  private readonly files = new Map<string, { content: string; sizeBytes: number; mtimeMs: number }>();
  private clock = 1;

  private tick(): number {
    return this.clock++;
  }

  exists(path: string): boolean {
    return this.files.has(path);
  }

  write(path: string, content: string): void {
    const sizeBytes = Buffer.byteLength(content, "utf-8");
    this.files.set(path, { content, sizeBytes, mtimeMs: this.tick() });
  }

  read(path: string): string | null {
    const f = this.files.get(path);
    return f ? f.content : null;
  }

  delete(path: string): void {
    this.files.delete(path);
  }

  setMtime(path: string, mtimeMs: number): void {
    const f = this.files.get(path);
    if (f) f.mtimeMs = mtimeMs;
  }

  /** Advance the store's logical clock and stamp `path` — mirrors File.SetLastWriteTimeUtc(now). */
  touchNow(path: string): void {
    const f = this.files.get(path);
    if (f) f.mtimeMs = this.tick();
  }

  list(root: string, suffix: string): Array<{ path: string; sizeBytes: number; mtimeMs: number }> {
    const prefix = root.endsWith("/") ? root : root + "/";
    const out: Array<{ path: string; sizeBytes: number; mtimeMs: number }> = [];
    for (const [path, f] of this.files) {
      if (path.startsWith(prefix) && path.endsWith(suffix)) {
        out.push({ path, sizeBytes: f.sizeBytes, mtimeMs: f.mtimeMs });
      }
    }
    return out;
  }
}

/**
 * Manages an on-disk cache of warm sessions keyed by the hash of
 * (modelId, systemPrompt). Ported from CircleAI.Inference.PrefixCacheService.
 */
export class PrefixCacheService {
  private static _default: PrefixCacheService | undefined;

  private readonly root: string;
  private readonly store: IPrefixCacheStore;

  /** The default per-app instance rooted at an in-memory cache directory. */
  static default(): PrefixCacheService {
    if (!PrefixCacheService._default) {
      PrefixCacheService._default = new PrefixCacheService(
        PrefixCacheService.defaultRoot(),
        new InMemoryPrefixCacheStore(),
      );
    }
    return PrefixCacheService._default;
  }

  /**
   * Construct a cache service rooted at `root` with the given backing store
   * (defaults to an in-memory store).
   */
  constructor(root: string, store?: IPrefixCacheStore) {
    if (!root || root.trim().length === 0) throw new Error("root is required.");
    this.root = root;
    this.store = store ?? new InMemoryPrefixCacheStore();
  }

  /**
   * Compute the cache key for a (modelId, systemPrompt) pair. Returns null when
   * systemPrompt is null/empty — nothing to cache without a system prompt.
   * Ported verbatim from PrefixCacheService.KeyFor.
   */
  static keyFor(modelId: string, systemPrompt: string | null | undefined): string | null {
    if (!modelId || modelId.trim().length === 0) return null;
    if (!systemPrompt || systemPrompt.length === 0) return null;

    const modelHash = sha256Hex(modelId);
    const systemHash = sha256Hex(systemPrompt);
    // First 16 hex chars per component — matches C#.
    return `${modelHash.substring(0, 16)}_${systemHash.substring(0, 16)}`;
  }

  /** Returns the cache path for `key`. Matches PathFor. */
  pathFor(key: string): string {
    return joinPath(this.root, `${key}.session`);
  }

  /** true when a cached entry exists for `key`. Matches HasEntryAsync. */
  async hasEntry(key: string): Promise<boolean> {
    return this.store.exists(this.pathFor(key));
  }

  /**
   * Write a session entry for `key` (deterministic marker content). Not present
   * in C# as a public method — the native generator wrote via MNN — but here it
   * is the concrete "snapshot" op the deterministic generator performs.
   */
  async writeEntry(key: string): Promise<void> {
    this.store.write(this.pathFor(key), `session:${key}`);
  }

  /** Write raw content at an arbitrary path (session save/load markers). */
  async writeRaw(path: string, content: string): Promise<void> {
    this.store.write(path, content);
  }

  /** Read raw content at an arbitrary path, or null. */
  async readRaw(path: string): Promise<string | null> {
    return this.store.read(path);
  }

  /**
   * Touch the entry's mtime so LRU eviction treats it as recently used. Called
   * after a successful load. Matches Touch.
   */
  touch(key: string): void {
    const path = this.pathFor(key);
    if (this.store.exists(path)) {
      if (this.store instanceof InMemoryPrefixCacheStore) this.store.touchNow(path);
      else this.store.setMtime(path, Date.now());
    }
  }

  /**
   * Evict oldest entries until the directory is under 500 MB. Called after
   * every successful save. Ported from EvictIfNeededAsync — oldest (smallest
   * mtime) first.
   */
  async evictIfNeeded(): Promise<void> {
    const files = this.store
      .list(this.root, ".session")
      .sort((a, b) => a.mtimeMs - b.mtimeMs);

    let total = files.reduce((sum, f) => sum + f.sizeBytes, 0);
    let i = 0;
    while (total > CAP_BYTES && i < files.length) {
      const f = files[i++]!;
      total -= f.sizeBytes;
      this.store.delete(f.path);
    }
  }

  private static defaultRoot(): string {
    // Windows: %LOCALAPPDATA%/CircleAI/prefix-cache
    // Unix-like: ~/.circleai/prefix-cache
    const local = process.env.LOCALAPPDATA;
    if (local && local.trim().length > 0) {
      return joinPath(local, joinPath("CircleAI", "prefix-cache"));
    }
    const home = process.env.HOME ?? process.env.USERPROFILE ?? ".";
    return joinPath(home, joinPath(".circleai", "prefix-cache"));
  }
}

// ── helpers ──────────────────────────────────────────────────────────────────

function sha256Hex(input: string): string {
  return createHash("sha256").update(input, "utf-8").digest("hex");
}

function joinPath(a: string, b: string): string {
  const left = a.endsWith("/") || a.endsWith("\\") ? a.slice(0, -1) : a;
  return `${left}/${b}`;
}

File: rust/src/inference/prefix_cache.rs
//! prefix_cache.rs
//!
//! RT-06: cross-session prefix cache. Ported from
//! `CircleAI.Inference/PrefixCacheService.cs`.
//!
//! Snapshot the model's warm KV state once per (modelId, systemPrompt) pair,
//! reload it on the next chat with the same pair, and skip the prefill entirely.
//!
//! The C# service owns only the *indexing* — the on-disk format is MNN's. Per
//! the no-real-IO porting brief, this Rust port keys the same way (byte-exact
//! [`PrefixCacheService::key_for`]) but stores snapshot payloads + their
//! last-used ordinal in memory. Eviction keeps the cache under a 500 MB cap by
//! dropping least-recently-used entries first — the same LRU-by-mtime policy the
//! C# `EvictIfNeededAsync` runs, with a monotonically increasing "touch"
//! ordinal standing in for the file mtime.

use std::collections::HashMap;

use crate::memory::multimodal::compute_sha256;

const CAP_BYTES: u64 = 500 * 1024 * 1024; // 500 MB

/// One cached warm-session entry.
#[derive(Debug, Clone)]
struct Entry {
    /// Opaque snapshot payload (the KV snapshot bytes MNN would have written).
    payload: Vec<u8>,
    /// Monotonic last-used ordinal; higher = more recently used.
    touched: u64,
}

/// Manages a cache of "warm" model sessions keyed by the hash of (modelId,
/// systemPrompt). Generators that opt in via `GenerationOptions.use_prefix_cache`
/// consult this service before resetting the model handle for a new
/// conversation. Thread-unsafe by itself — wrap in a `Mutex`/`RwLock` when
/// shared across threads (the C# service serialises via a semaphore).
#[derive(Debug, Default)]
pub struct PrefixCacheService {
    entries: HashMap<String, Entry>,
    clock: u64,
}

impl PrefixCacheService {
    /// Constructs an empty cache service.
    pub fn new() -> Self {
        Self {
            entries: HashMap::new(),
            clock: 0,
        }
    }

    /// Compute the cache key for a (modelId, systemPrompt) pair. Returns `None`
    /// when `model_id` is blank or `system_prompt` is `None`/empty — there is
    /// nothing to cache without a system prompt to key against.
    ///
    /// Byte-exact with the C#: `sha256(model_id)[..16] + "_" + sha256(system)[..16]`.
    pub fn key_for(model_id: &str, system_prompt: Option<&str>) -> Option<String> {
        if model_id.trim().is_empty() {
            return None;
        }
        let system = match system_prompt {
            Some(s) if !s.is_empty() => s,
            _ => return None,
        };

        let model_hash = compute_sha256(model_id.as_bytes());
        let system_hash = compute_sha256(system.as_bytes());
        // First 16 hex chars per component — collision-free at any single
        // device's cache scale, much shorter on disk.
        Some(format!("{}_{}", &model_hash[..16], &system_hash[..16]))
    }

    /// `true` when a cached entry exists for `key`.
    pub fn has_entry(&self, key: &str) -> bool {
        self.entries.contains_key(key)
    }

    /// Number of currently cached entries. Diagnostics only.
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    /// `true` when the cache holds no entries.
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Store a warm-session snapshot for `key`, then bound the cache. Mirrors
    /// the C# `SaveSession` → `EvictIfNeededAsync` sequence: the write always
    /// happens, then eviction keeps the total under the cap.
    pub fn save(&mut self, key: &str, payload: impl Into<Vec<u8>>) {
        self.clock += 1;
        let touched = self.clock;
        self.entries.insert(
            key.to_string(),
            Entry {
                payload: payload.into(),
                touched,
            },
        );
        self.evict_if_needed();
    }

    /// Load a previously-saved snapshot, touching it so LRU eviction treats it
    /// as recently used (mirrors the C# `Touch` after a successful load).
    /// Returns `None` when the key is absent.
    pub fn load(&mut self, key: &str) -> Option<Vec<u8>> {
        self.clock += 1;
        let now = self.clock;
        match self.entries.get_mut(key) {
            Some(e) => {
                e.touched = now;
                Some(e.payload.clone())
            }
            None => None,
        }
    }

    /// Touch the entry so LRU eviction treats it as recently used. No-op when
    /// the key is absent. Mirrors the C# `Touch(key)`.
    pub fn touch(&mut self, key: &str) {
        self.clock += 1;
        let now = self.clock;
        if let Some(e) = self.entries.get_mut(key) {
            e.touched = now;
        }
    }

    /// Total bytes currently held across all cached snapshots.
    pub fn total_bytes(&self) -> u64 {
        self.entries.values().map(|e| e.payload.len() as u64).sum()
    }

    /// Evict least-recently-used entries until the cache is under the 500 MB
    /// cap. Called after every `save`. Best-effort and idempotent.
    pub fn evict_if_needed(&mut self) {
        let mut total = self.total_bytes();
        if total <= CAP_BYTES {
            return;
        }
        // Order keys oldest-first (lowest touch ordinal), evicting until under.
        let mut ordered: Vec<(String, u64, u64)> = self
            .entries
            .iter()
            .map(|(k, e)| (k.clone(), e.touched, e.payload.len() as u64))
            .collect();
        ordered.sort_by_key(|(_, touched, _)| *touched);

        for (key, _touched, size) in ordered {
            if total <= CAP_BYTES {
                break;
            }
            self.entries.remove(&key);
            total -= size;
        }
    }
}
